src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,370 @@
+	Prefix string         // Prefix is prepended to node keys stored in etcd to avoid naming conflicts.
+	Logger *logrus.Logger // Logger is used to output log messages if defined.
+
+	Cancel context.CancelFunc // Cancel can be used bya caller to stop Ring subroutines.
+	Ctx    context.Context    // Ctx is the context used to create the Ring.
+
+	node    node     // node represents the member instance in a Ring that is local to the current process.
+	members []Member // members is a string array of all node Ids in the Ring sorted by hashFn.
+
+	hashFn func([]byte) uint32 // hashFn will be called to determine the hash for a given key.
+
+}
+
+// node is a local Member of a Ring. Upon creation, a node requests a lease from etcd and continuously refreshes it
+// until it is shutdown.
+type node struct {
+	Member
+	Locks sync.Map // Locks is a thread-safe map of lockInfos associated with a node.
+}
+
+// Member contains metadata about members in a Ring.
+type Member struct {
+	Id   string // Id is the name of a members.
+	Hash uint32 // Hash is the output of the ring's hashFn(Id).
+}
+
+// lockInfo instances are created for keys by a ring's node.
+// A context lockInfo.Ctx is stored so the proper context is used when unlocking lockInfo.Lock
+type lockInfo struct {
+	Lock dlock.DLock     // Lock is a dlock that locks a key with a prefix in etcd.
+	Ctx  context.Context // Ctx is the context returned by Lock.
+}
+
+// New is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func New(ctx context.Context, prefix string, client *etcd.Client, logger *logrus.Logger, id ...string) (*Ring, error) {
+	if id == nil {
+		id = []string{uuid.New()}
+	}
+
+	ring := defaultRing(ctx, prefix, client, logger)
+	if err := ring.start(id[0]); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	return ring, nil
+}
+
+// GetIndex returns the index of the Member associated with key. A key belongs to the first Member whose hash is
+// greater than the key. If no Members are greater, then the Ring wraps around to the first Member.
+func (ring *Ring) GetIndex(key string) int {
+	keyHash := ring.hashFn([]byte(key))
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= keyHash
+	})
+
+	// wrap around to the first node in the Ring.
+	if index >= len(ring.members) {
+		index = 0
+	}
+
+	return index
+}
+
+// Get is a convenience wrapper around GetIndex that returns the Member that associates to a key.
+func (ring *Ring) Get(key string) Member {
+	return ring.members[ring.GetIndex(key)]
+}
+
+// Lock attempts to lock a key in etcd using consistent hashing. If key does not belong to the ring's node,
+// the call blocks until context is cancelled or expires.
+func (ring *Ring) Lock(ctx context.Context, key string) (context.Context, error) {
+	ticker := time.NewTicker(time.Second * 5)
```

Comment:
I would add a TODO comment here to make this event driven. This polling strategy might be fine for now, but it may be problematic at scale.